Libs/MRML/Core/vtkMRMLBitStreamVolumeNode.cxx

```diff
@@ -0,0 +1,247 @@
+{
+  destString.resize(srcString.length());
+  char * array = new char[srcString.length()];
+  memcpy(array, srcString.c_str(), srcString.length());
+  destString.assign(array, srcString.length());
+}
+
+void vtkMRMLBitStreamVolumeNode::SetFrameMessage(std::string& buffer)
+{
+  this->CopySrcStringToDestString(buffer, this->FrameMessage);
+  this->FrameMessageValid = true;
+};
+
+std::string vtkMRMLBitStreamVolumeNode::GetFrameMessage()
+{
+  std::string returnMSG(this->FrameMessage);
+  return returnMSG;
+};
+
+void vtkMRMLBitStreamVolumeNode::SetKeyFrameMessage(std::string& buffer)
+{
+  this->CopySrcStringToDestString(buffer, this->KeyFrameMessage);
+};
+
+std::string vtkMRMLBitStreamVolumeNode::GetKeyFrameMessage()
+{
+  std::string returnMSG(this->KeyFrameMessage);
+  return returnMSG;
+};
+
+//---------------------------------------------------------------------------
+int vtkMRMLBitStreamVolumeNode::DecodeFrameMessage(vtkMRMLCompressionDeviceNode::ContentData* deviceContent)
+{
+  if (this->CompressionDevice == NULL)
+  {
+    vtkWarningMacro("No compression device available, use the ObserveOutsideCompressionDevice() to set up the compression device.")
+    return 0;
+  }
+  if (this->GetImageData() != this->CompressionDevice->GetContent()->image)
+  {
+    this->SetAndObserveImageData(this->CompressionDevice->GetContent()->image);
+  }
+  if(!this->GetKeyFrameDecoded())
+    {
+    this->SetKeyFrameDecoded(true);
+    this->SetKeyFrameMessage(deviceContent->keyFrameMessage);
+    int decodingStatus = this->CompressionDevice->UncompressedDataFromBitStream(deviceContent->keyFrameMessage, true);
+    if (decodingStatus == 0)
+      {
+      vtkWarningMacro("key frame decoding failed.")
+      return 0;
+      }
+    }
+  std::string bitStreamData = std::string(deviceContent->frameMessage);
+  if (this->CompressionDevice->UncompressedDataFromBitStream(bitStreamData, true))
+    {
+    this->Modified();
+    return 1;
+    }
+  return 0;
+}
+
+//---------------------------------------------------------------------------
+int vtkMRMLBitStreamVolumeNode::EncodeImageData()
+{
+  if (this->CompressionDevice == NULL)
+    {
+    vtkWarningMacro("No compression device available, use the ObserveOutsideCompressionDevice() to set up the compression device.")
+    return 0;
+    }
+  if (this->GetImageData() != this->CompressionDevice->GetContent()->image)
+    {
+    this->CompressionDevice->SetContentImage(this->GetImageData());
+    }
+  std::string compressedStream = this->CompressionDevice->GetCompressedBitStreamFromData();
+  if (compressedStream.size()>0)
+    {
+    this->SetFrameMessage(compressedStream);
+    if(!this->GetKeyFrameReceived())
+      {
+      this->SetKeyFrameReceived(true);
+      this->SetKeyFrameMessage(compressedStream);
+      }
+      return 1;
+    }
+  return 0;
+}
+
+std::string vtkMRMLBitStreamVolumeNode::GetCodecType()
```

Comment:
Probably the interface should have a CodecName property (simple string), that would be saved in MRML scene; and a Codec property (VTK class based vtkStreamingVideoCodec interface class) that would be set by the Volumes logic.

Reply:
Addressed, using CodecName and CodecDeviceType